servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/HttpRequestAutoDrainingServiceFilter.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright © 2025 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.http.utils;
+
+import io.servicetalk.concurrent.PublisherSource;
+import io.servicetalk.concurrent.api.Single;
+import io.servicetalk.concurrent.api.TerminalSignalConsumer;
+import io.servicetalk.concurrent.internal.CancelImmediatelySubscriber;
+import io.servicetalk.http.api.HttpExecutionStrategy;
+import io.servicetalk.http.api.HttpRequest;
+import io.servicetalk.http.api.HttpServiceContext;
+import io.servicetalk.http.api.StreamingHttpRequest;
+import io.servicetalk.http.api.StreamingHttpResponse;
+import io.servicetalk.http.api.StreamingHttpResponseFactory;
+import io.servicetalk.http.api.StreamingHttpService;
+import io.servicetalk.http.api.StreamingHttpServiceFilter;
+import io.servicetalk.http.api.StreamingHttpServiceFilterFactory;
+
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.concurrent.api.SourceAdapters.toSource;
+import static io.servicetalk.http.api.HttpExecutionStrategies.offloadNone;
+
+/**
+ * Configure automatic consumption of request {@link StreamingHttpRequest#payloadBody() payload body} when it is not
+ * consumed by the service.
+ * <p>
+ * For <a href="https://tools.ietf.org/html/rfc7230#section-6.3">persistent HTTP connections</a> it is required to
+ * eventually consume the entire request payload to enable reading of the next request. This is required because
+ * requests are pipelined for HTTP/1.1, so if the previous request is not completely read, next request can not be
+ * read from the socket. For cases when there is a possibility that user may forget to consume request payload,
+ * ServiceTalk automatically consumes request payload body.
+ * An example of guaranteed consumption are {@link HttpRequest non-streaming APIs}.
+ *
+ * @see io.servicetalk.http.api.HttpServerBuilder#drainRequestPayloadBody(boolean)
+ */
+public final class HttpRequestAutoDrainingServiceFilter implements StreamingHttpServiceFilterFactory {
```

Comment:
I move this to http-util, which feels like the right spot for it. Opinions welcome.